### ISSUE TITLE
Add source table ID parameter to BigQuery cell magic

### ISSUE THREAD

**Issue by plamut** [googleapis/python-bigquery#62]
Following a [comment](https://github.com/googleapis/python-bigquery/issues/53#issuecomment-600252957) in another thread, a user expressed the need for dynamically specifying the source table ID.

 **Describe the solution you'd like**
The user wants to templatize Jupyter notebooks, and they need a dynamic way of specifying the source table ID, similarly to the existing `--destination_table` parameter. The existing feature that allows the table ID to be specified in the cell body is not sufficient.

**Describe alternatives you've considered**
The first attempt was to use query parameters, but these can only be used for query data, and not for table, column, etc. identifiers ([documentation](https://cloud.google.com/bigquery/docs/parameterized-queries#running_parameterized_queries)).

There is a plan to add support for "raw" queries, but constructing these dynamically is a potential security risk, because no automatic parameter escaping takes place. Additionally, it will take some time to implement this feature on the backend, and the release date is not even known yet.

**Comment by plamut** [comment]
@julioyildo Does the description correctly reflect what you would need and would actually solve your use case? Would you also need anything else on top of the new parameter?

**Comment by julioyildo** [comment]
@plamut Thank you, that sums up our needs. :) 

I would add : 
> `--source_tables`  having the ability to enter multiple Input Tables (as `--params`) would be even better.

That will be all. Thank you very much!

**Comment by plamut** [comment]
@julioyildo Just as a clarification, what is the expected semantics/outcome in the two use cases below?

It is important that the new feature is generally useful and not tied to a few specific use cases. It should be something applicable to the wider audience, thus the typical use cases, as well as corner cases, need to be clearly defined.

### Multiple sources and a query
```
%%bigquery --source-tables table_1, table_2 --destination_table table_3
SELECT foo FROM table_4
```
What exactly should be written to `table_3` and how? There are multiple sources specified, as well as a query in the cell body.

### A single source and a table ID
```
%%bigquery --source-tables table_1
my_project.my_dataset.table_2
```
A source table and a table ID in cell body are both given. What should happen, which dataset should be fetched?

---

By the way, would it also be possible to achieve the same with a few short code snippets using a BigQuery client instance? These could be different for every customer with a few configurable variables. Or did I misunderstood how you envision the notebook templatization?

The thing to keep in mind here is the cell magic is primarily meant just as a convenience helper for quickly fetching data, but it cannot cover all use cases.

Thanks for the clarifications in advance!

**Comment by julioyildo** [comment]
@plamut Sure ! 

### Multiple sources and a query

`--destination_table` : this param works well. It saves the output of the query to a new BigQuery table. In the example below `table_3` will be created.

`--source_tables` : this parameter could work in the same way as `--params` by defining flag and  values (=tables_ids)

```
%%bigquery --source-tables {"my_tbal_1":"project_1.dataset_1.table_1","my_tbal_2":"project_2.dataset_2.table_2"} 
           --destination_table project3.dataset_3.table_3

SELECT *
FROM @my_tbal_1
UNION ALL
SELECT *
FROM @my_tbal_2
```
- **Here is a quick notebook use case :**
https://drive.google.com/file/d/1aTTRKyjB7Jiq-eNPWEA3FunAiYY4pbgb/view



### A single source and a table ID

`--source_table` values should be considered only if the declared flags are used in the body cell.
In this case, even if `source_tables` is declared, the result of the query would not impacted and `my_project.my_dataset.table_2`  will returned.

The correct response would be : 
```
%%bigquery 
my_project.my_dataset.table_2
```

-----------------------

### Here is a code snippet example to be added to  `google.cloud.bigquery.magics`

This is a quick example (not optimized) of code that could address the problem.


- **Function to be added in** `magics.py`
```
def _handle_source_tables(query, source_tables):
     """Runs a query while printing status updates

    Args:

        query (str):
            SQL query to be executed. Defaults to the standard SQL dialect.

        source_tables (dict):
            Tables names dictionnary to be injected in query
            
    Returns:
         query (str):
             New SQL query 
      """

    try:
        if isinstance(source_tables, dict):
            reg = re.split(r'(@\w+)', query)
            print('loop')
            for key in source_tables:
                key_to_match = '@'+key
                key_id = reg.index(key_to_match)
                reg[key_id] = source_tables[key]
            
            query = ''.join(reg)
            return query

    except Exception:
        raise
        
        
```
- **Magic Argument to be added** :

```
@magic_arguments.argument(
    "--source_tables",
    nargs="+",
    default=None,
    help=(
        "Parameters to format the query string. If present, the --source_table "
        "in the format {'flag_name': 'table_id'} (ex. {'table_1': 'project_name.dataset_name.table_name'}), "
        "or a reference to a dictionary in the same format. The dictionary "
        "reference can be made by including a '$' before the variable "
        "name (ex. $my_table_dict_var)."
    ),
)
```


-  **Finally a few lines to be added in**  `def _cell_magic()` :
```

def _cell_magic(line, query):

    args = magic_arguments.parse_argstring(_cell_magic, line)

    params = []
    if args.params is not None:
        try:
            params = _helpers.to_query_parameters(
                ast.literal_eval("".join(args.params))
            )
        except Exception:
            raise SyntaxError(
                "--params is not a correctly formatted JSON string or a JSON "
                "serializable dictionary"
            )



   ###################
    # --source_tables:
    ##################
    if args.source_tables is not None:
        try:
           source_tables =ast.literal_eval("".join(args.source_tables))
           query =_handle_source_tables(query, source_tables)

        except Exception as ex:
            _handle_error(ex, args.source_tables)
            return 
   
     ###################
    # END --source_tables:
    ##################






    #  REST OF CODE UNCHANGED...#
    project = args.project or context.project
    client = bigquery.Client(
        project=project,
        credentials=context.credentials,
        default_query_job_config=context.default_query_job_config,
        client_info=client_info.ClientInfo(user_agent=IPYTHON_USER_AGENT),
    )
```

**Comment by plamut** [comment]
@julioyildo Thanks, this is useful!

Dynamic table names are therefore a sort of a query pre-processing macro, replacing `@table_name` placeholders with the actual names before submitting the query to the backend.

The `@` sign currently does not appear to be a part of the [standard SQL syntax](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax), and could actually be used, albeit at a risk of a future incompatibility, if the supported syntax is extended on the backend.

@shollyman What are your thoughts on this? The feature itself, as proposed, would not be too difficult to add, but we need to assure that it does not burden future development.

**Comment by shollyman** [comment]
So, something that's recently arrived in the SQL language is EXECUTE IMMEDIATE, which allows arbitrary text to be composed and run as a query.  This gets around the constraints around existing query parameters.  See https://cloud.google.com/bigquery/docs/reference/standard-sql/scripting#execute_immediate for the reference information.

Presumably with this usage, you could do the same kind of parameter binding without the need for more explicit magics being added.

**Comment by plamut** [comment]
@julioyildo Does the EXECUTE IMMEDIATE addition help with your use case? Is this feature request still relevant in the light of the former?

**Comment by julioyildo** [comment]
Hi @shollyman & @plamut,

First of all, thank you for your answers, and sorry for my response time.

After having tested the EXECUTE IMMEDIATE Solution here are our findings: 

- **EXECUTE IMMEDIATE seems not compatible with Python Bigquery Magics library. During our tests on Jupyter Notebook, we raised a constant error indicating:** 
  "400 Not found: Dataset was not found in location US at [1:1]".

    However, our datasets are located in the right areas.

- **It's not possible to parameter Table Names and Column Names.** Without that, the feature can't fit our needs; templating our notebooks.

- The feature is working with Stringified SQL, **which causing User Readability difficulties** in case of longs scripts.

Meanwhile, we will carefully monitor the evolution of this feature and will use it when it meets our expectations. 

----

Currently, we already implemented the feature above, it's working correctly and the results are very interesting for our use cases.

We still think that this feature could be a good complement to the Python-Bigquery library, and an interesting asset for the Bigquery Magics users. 

Thank you
